Refactor and clean up jwxml

josePhoenix opened the pull request:
This should not change any functionality, but rearranges the existing code into a `jwxml` package with `siaf` and `mirrors` subpackages. This will make it easier to subtree into an existing package, bundle a test suite, and other such things.

Commit:
Convert jwxml from module to package

Commit:
Smoke test SIAF in new refactored arrangement

Commit:
Remove redundant iterchildren definition

Commit:
PyCharm auto-format siaf.py

Commit:
Remove dead code from siaf.py

Commit:
Merge branch 'master' of https://github.com/mperrin/jwxml

Commit:
Typo fix

Commit:
Package the SIAF data in the jwxml package

Commit:
Fix bug in iterchildren wrapper

mperrin commented on jwxml/mirrors.py:
```diff
@@ -0,0 +1,141 @@
+
+        self.units = dict()
+        self.moves = dict()
+        for move in iterchildren(xmlnode):
```
I think this would fail right now because the `iterchildren` wrapper is only defined in `siaf.py` and is invisible in `mirrors.py`. Maybe we should stick that in a `util.py`?

josePhoenix replied:
Noticed this and fixed it locally, but haven't yet pushed!

mperrin replied:
:thumbsup:

mperrin commented on jwxml/siaf.py:
```diff
@@ -0,0 +1,739 @@
+
+
+class Test_SIAF(unittest.TestCase):
+    def assertAlmostEqualTwo(self, tuple1, tuple2):
```
Should we relocate this to `tests/test_siaf.py`? It will also need updating to fix the current hard coded filename for a six year old SIAF instance. But that can be a separate PR after this one of course.

josePhoenix replied:
I'm working this up along with the inclusion of SIAF snapshots. Agreed that should be a separate PR. Hopefully I haven't tangled the history up too much.

mperrin replied:
:thumbsup:

Commit:
Move iterchildren helpers to 'utils', make constants.py with data paths

Commit:
(not yet working) Conversion of SIAF tests from unittest to py.test

josePhoenix commented:
Well, in updating the test cases, I seem to have found a regression.

```
_____________________________________________ test_inverse_transforms ______________________________________________

nircam_siaf = <jwxml.siaf.SIAF object at 0x116500390>

    def test_inverse_transforms(nircam_siaf):
        nca = nircam_siaf['NRCA4_FULL']

        assert pair_almost_equal(nca.Det2Sci(*nca.Sci2Det(1020., 1020)), (1020., 1020))
        assert pair_almost_equal(nca.Sci2Det(*nca.Det2Sci(1020., 1020)), (1020., 1020))

        assert pair_almost_equal(nca.Tel2Idl(*nca.Idl2Tel(10., 10)), (10., 10))
        assert pair_almost_equal(nca.Idl2Tel(*nca.Tel2Idl(10., 10)), (10., 10))

>       assert pair_almost_equal(nca.Tel2Sci(*nca.Sci2Tel(10., 10)), (10., 10))

jwxml/tests/test_siaf.py:58:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

left = (9.1403015993034842, 11.031090255990307), right = (10.0, 10), atol = 1e-06

    def pair_almost_equal(left, right, atol=1e-6):
        """Compare whether two 2-tuples (pairs) of numeric values have an
        absolute difference of < `atol`"""
        left_1, left_2 = left
        right_1, right_2 = right
>       assert abs(left_1 - right_1) < atol, "first elements differ by more than {}".format(atol)
E       AssertionError: first elements differ by more than 1e-06
E       assert 0.85969840069651582 < 1e-06
E        +  where 0.85969840069651582 = abs((9.1403015993034842 - 10.0))
```

Commit:
Complete SIAF Aperture test conversion to py.test

josePhoenix commented:
One item for discussion: I have structured the package + data this way:

```
jwxml/
    siaf.py
    mirrors.py
    data/
        PRDDEVSOC-D-016/
            NIRCam_SIAF.xml
            ...
    tests/
        test_siaf.py
```

Then `jwxml.PRD_VERSION` says `'PRDDEVSOC-D-016'` and `jwxml.PRD_DATA_ROOT` says `<install path>/jwxml/data/PRDDEVSOC-D-016/`. My thought was that we could keep the old PRD versions around, and it would always be clear what versions you had and what you were updating because you'd never be replacing files, just adding new ones.

Of course, that was a bit naive considering the XML files weigh in at 2MB per delivery. Disk is cheap and all, but assuming ~5 more deliveries this size before the telescope is commissioned... that makes jwxml a 12 MB dependency, and we'll be dragging around outdated info until the end of time.

I should probably talk this over with Matt Lallo, since in addition to the PRD updates there will also be `UseAfter`-type conditions that may require having multiple SIAFs around to completely specify a position mapping at every point in the lifetime of the instrument.

mperrin commented:
Of course given the way git works, you have to drag around all past versions of files forever too, so it doesn't reduce the disk usage any to overwrite files.  I'm inclined to not worry about that -- after all, webbpsf + poppy are already 100 MB (mostly dominated by PNGs in docs and notebooks?) So I think you are prematurely optimizing to worry about 2 MB here and there.

mperrin commented:
And meanwhile it's both annoying and useful that you found that test case failure. :-)

josePhoenix commented:
Well, the thing about git dragging around old versions is true for binary files, but in theory text-based formats can be stored as deltas. In practice, I think the SIAF generation tools reorder things arbitrarily, so the deltas between two PRD versions would be meaningless.

Just checked that the test failure isn't from my refactoring, and I can reproduce with mperrin/jwxml@master

josePhoenix commented:
Another pro to keeping siaf.py with the old SIAFs: git bisecting for a regression would be much easier. I'm not sure we have a version of the code and a SIAF that passes that test at the moment. 😞

Commit:
SIAF convert() should require frame_from and frame_to arguments

josePhoenix commented:
@mperrin any objections to merging this now? (I don't have push access to this repo...)

mperrin commented:
Are we content that there are no code issues related to that failed unit test you uncovered (in #6)?  If so then I will merge. But I had the impression that Matt L is somewhat less comfortable with the round-tripping error being "as expected" than Colin C is.

mperrin commented on jwxml/siaf.py:
```diff
@@ -0,0 +1,666 @@
+    from webbpsf import utils
+except ImportError:
+    utils = None
+
```
I think this whole block here where it tries to import utils from webbpsf is unnecessary vestigial code - it doesn't look like it gets used anywhere.

mperrin commented:
If we think that test is bogus, then we should modify that test to either xfail, or change the test coordinates to ones that work better or adjust the tolerance. I'd like to get all the unit tests passing again while we're working on this, and then I can set up things on Travis.

Commit:
Remove vestigial webbpsf import

Commit:
Make siaf.py importable even when data are missing

josePhoenix commented:
I have checked that mperrin/jwxml `master` exhibits the same issue with the forward/reverse transform agreement. So, this is not an issue introduced in this PR, and likely requires a change to the SIAF for improved agreement. (I believe I checked every equation against the tech memo when Chris Stark and I were trying to debug transforms in the Target Visibility Tool, so I'm doubtful that there's a math error.)

I removed the vestigial webbpsf import, and I moved a check for the presence of the data files into the step where they are loaded. I guess I was thinking "oh, if the data files are missing, that's a packaging/installation problem and should be caught early", but in practice that wouldn't do anything to help users with installation problems and would annoy users who know why that check is failing.

josePhoenix commented:
I'm going to branch off this and make a dead simple setup.py for this. I think the `git subtree` idea was not a win for maintainability after all.

mperrin commented:
> I'm going to branch off this and make a dead simple setup.py for this. I think the git subtree idea was not a win for maintainability after all.

Can you say more what you're thinking? How do you envision including this within webbpsf and the target vis tool?

josePhoenix commented:
Once there's a `setup.py` in place, we can use this (the same way we've been told to depend on pysynphot by SSB): https://setuptools.readthedocs.io/en/latest/setuptools.html#dependencies-that-aren-t-in-pypi

(The only cost I'm aware of is that users must have git installed, but it's installed by default on most systems.)

josePhoenix commented:
Ugh, it turns out this is something upstream is not keen on: http://stackoverflow.com/questions/12518499/pip-ignores-dependency-links-in-setup-py

I don't know. Putting it on PyPI isn't a ton of work. There's also the git subtree technique, which would put a copy each in webbpsf, jwst_visibility, and siaf_viewer. There's also git submodules, which are perhaps easier to peg to a version, but more unintuitive because they don't appear when cloned.

mperrin commented:
What's your argument against subtree? I'm not keep on submodules given the potential for bafflement since they don't get cloned.  Subtree seems to have been working without too many headaches for the current version of jwxml. 

I'm not deeply opposed to putting it on pypi. It just seems like polluting the namespace needlessly to post such an obscure single-use package there.

josePhoenix commented:
The current version actually uses submodule (see https://github.com/mperrin/webbpsf/blob/master/.gitmodules).

To use subtree, I'd have to move the the contents of `jwxml` to the root of the repository alongside any docs or other files, such that the `webbpsf/jwxml/` subtree itself was an importable package. (Or else do hacks like `sys.path.insert(os.path.join(os.path.dirname(__file__), 'jwxml'), 1)` in `__init__.py`, which have the potential for surprises.)

mperrin commented:
whoops, I'd gotten confused which we were using already. hmm. 

Hacky manipulations of sys.path seem like they should be strongly avoided.

josePhoenix commented:
Interestingly, `python setup.py develop` will install dependencies listed in `dependency_links=`, but pip won't follow them.

There are enough empty/useless packages cluttering up PyPI that I think an actual useful one (albeit niche) is small potatoes.

mperrin commented:
side question - if we're going to stick it on pypi, should we pick a better name first? 'jwxml' was just my internal shorthand for a single utils file inside webbpsf, and I never gave it any particularly deep thought. Should we be more verbose with 'jwst_xml_utils' or something like that? 

Conversely, given the vast plethora of other uses of XML files throughout the project, it's sort of presumptuous to say "this is _the_ JWST XML package" because of course it's not any sort of project-level official thing, it's just our little swiss army knife. But 'jwst_ote_xml_utils' is even uglier.

josePhoenix commented:
I'm not sure we'll do a whole lot better. Underscores are mildly discouraged in package names, in favor of single "words", so I think jwxml is pretty close to optimal.
